Workflow Editor phase 2, part 1: Task Funnel

shaunanoordin opened the pull request:
## PR Overview

Part of: **Workflow Editor** (aka **Pages Editor**) project phase 2
Related issue: #7240
Staging branch URL: https://pr-7245.pfe-preview.zooniverse.org/lab/1982/workflows/editor/3711?env=staging

This PR is the first in a series of updates for the Workflow Editor project, phase 2. In this PR specifically, we're updating the UI to match Sean's new designs described in 7240 - a lot of the header elements have been redesigned.

Components:
- `<WorkflowHeader>` will now be split into multiple smaller components.
- **No functional changes expected,** only a lot of UI elements being rearranged.

Also:
- CSS classes will also be cleaned up to make the code more readable, and make more sense from an object/component-based POV (instead of a functional POV, so more _class="confirm-button"_ and less _class="button green-button flex-item fixed-position"_)
  - Experiment: we'll be using Stylus mixins to package common behaviour inside CSS classes.
  - ⚠️ This may or may not make the eventual migration to the FEM codebase easier or harder. Who knows! We've already earmarked the CSS/styling as one of the major things that needs to be addressed during migration.

### Status

~~WIP~~

Update: ready for review [(16 Jan 2025)](https://github.com/zooniverse/Panoptes-Front-End/pull/7245#issuecomment-2597230774)

seanmiller26 commented:
- [ ] Workflow title
I believe this should this be set to H1. Thoughts?

- [ ] Task/Workflow Settings tabs
The unselected tab should have a lighter font weight. Change to #5C5C5C for consistency. Hover state should match the new homepage, stats pages.

- [ ] Starting page dropdown
Removed! As per our last discussion, this isn't necessary anymore due to the new visual design and the decision that Pages should renumber themselves according to their position in the funnel.

- [ ] Status/Version
Adjust margin above and below to 10px. Also, checking in on this UX change from the issue: 
> UX improvement: increase timing of the loading icon to show for a longer duration after an update.

- [ ] Next task arrow
After linear tasks. As discussed, this element should be included in this PR. (Note, this should only show between tasks, but not at the end of the funnel)
<img width="300" alt="image" src="https://github.com/user-attachments/assets/ae2e5fd3-a79f-4a54-9351-12542e07053d" />
<img width="300" alt="image" src="https://github.com/user-attachments/assets/136b490b-30a3-4659-87ec-178e81d7b36d" />



- [ ] Add a Task button
Inspect in Figma for updated coloring and hover/click states. Match Figma font and icon size.

shaunanoordin commented:
Thanks for the detailed items, @seanmiller26 !

OK, here's an update as of 44b180a:

- Workflow Title: this is now h1. ✔️ 
  - This makes sense, the workflow title is the defining text of the page (unless we add something more clunkily descriptive like "page editor: editing (workflow title)") so
  - Side note: I also did a check and turns out the only other headers on the page are the h2s in the left-hand PFE menu. I'm surprised.
- Task/Workflow Settings tabs: updated ✔️ 
  - To be _specific,_ if a tab is selected, it has: color #404040, border-bottom 4px solid #D47811, font-weight 700 (as per Figma)
  - if a tab isn't selected, it has: color #5c5c5c (as per Github comment + Figma), border-bottom none, font-weight 500 (as per Figma)
  - 🆕 if a tab isn't selected AND is hovered on, it has: color #5c5c5c, border-bottom 4px solid **#000000**, font-weight 500
- Starting Page: removed ✔️ 
- Status/Version bit: margin top/bottom set to 10px... sort of. 🤨 
  - Here's the technical thing: there's some invisible space above the first task that serves as a "drag-and-drop _drop target._ I've highlighted the drag-and-drop _drop targets_ as green below, the workflow overview bar (the active/wf version bar) in red.
    <img width="849" alt="image" src="https://github.com/user-attachments/assets/738391ba-29fd-45a7-b482-8aba3c7c34c6" />
  - The workflow overview bar has margin 10px for the top, margin 0 elsewhere. The drop targets were previously 16px in height, now they're 10px in height.
  - Visually, this is fine. _Functionally,_ there's now an unexpected change where the drag & drop drop targets is a bit smaller than previously. Is this something users will notice? Maybe not, but it's worth highlighting that part of that workflow overview bar's (active / workflow version) margins are affecting functionality outside of its own presentation. 🤷 
- ⚠️ _UX improvement: increase timing of the loading icon to show for a longer duration after an update._ - this has NOT been implemented. ⚠️
  - **This may need to be its own side quest.** It's worth doing, but the effort needed is bigger than it seems.
  - The problem: the loading icon (component) is tied directly to the state of the workflow (data). The React code is set so that "if workflow is busy, show icon".
  - If we want to show the icon for longer, we need to change that logic to "if workflow is busy OR _was busy recently,_ show icon". The logic of "was busy recently" requires implementing a set of _timers_ in a React + network call context, which requires a fair amount of testing & planning to get right.
  - ~~EDIT: I'll also need to look into CSS animations to see if there's a simpler solution there.~~ UPDATE: still need some timers since the existence/non-existence of that loading icon has a small impact on page content.
- Next Task Arrows: implemented ✔️ 
  - **Hang on, shouldn't these be called next PAGE arrows?**
- Add a Task button: implemented ✔️ 

Did I miss anything? As usual, the preview's up at https://pr-7245.pfe-preview.zooniverse.org/lab/1982/workflows/editor/3711?env=staging

shaunanoordin commented:
## PR Update

Come to think of it, I'm now confident that this PR is ready for review. Let's go!

What to expect from this PR:

- Design: a lot of page elements have been rearranged and restyled.
  - See #7240 and [this update comment](https://github.com/zooniverse/Panoptes-Front-End/pull/7245#issuecomment-2590587517) for design specifics.
- Functionality: functionally, the Pages Editor should be mostly the same.
  - Exception: you can no longer set the staring page via a dropdown. The rule is, the first page in the Task funnel is the starting page.
- Code: the CSS code is in the process of being refactored to make it more readable to us.
  - Verbose "functional/atomic CSS class naming" like _class="button green-button flex-item fixed-position"_ has been replaced with more "descriptive CSS class naming" like _class="confirm-button"_
  - A lot of earlier decisions no longer make sense, e.g. `$sizeS = 8px; $sizeSM = 12px; etc` fixed sizes don't work when the design is being fine-tuned with a more fluid spectrum of sizes from 8px, 10px, 12px, 14px, etc.

Note that a number of obvious improvements (e.g. page numbers) are queued for a separate PR, and that will follow issue #7241

### Testing

For @seanmiller26: keep on reviewin' the way you've been reviewin' 👍

For devs:

- View the staging workflow: https://pr-7245.pfe-preview.zooniverse.org/lab/1982/workflows/editor/3711?env=staging
  - This is safe to edit, as long as nobody else is testing the same time as you.
- There should be **no major** functional changes.
  - On the Tasks page, you should be able to
    - [ ] (1) create Tasks,
    - [ ] (2) edit Pages,
    - [ ] (3) edit Tasks,
    - [ ] (4) rearrange Pages,
    - [ ] (5) delete Pages, and
    - [ ] (6) delete Tasks, as per usual.
    - NOTE: minor functional change - you can no longer set the starting page via the dropdown, since that has been removed.
  - [ ] On the Workflow Settings page, you should be able to change any setting, e.g. workflow title.
  - General sanity checks:
    - [ ] the Pages Editor should present to you, the user, all the information on the workflow that you'd want to know.
    - [ ] accessibility shouldn't be worse than prior to this PR.
- To view advanced options: https://pr-7245.pfe-preview.zooniverse.org/lab/1982/workflows/editor/3711?env=staging&advanced=1 (useful for resetting the workflow or quickly creating a standard test workflow)

### Status

Ready for review

shaunanoordin commented:
Derp, minor note: I only JUST pushed commit ab27a4d which added the Workflow Tab hover code. That was sitting on my local dev for the past day. 🤦

seanmiller26 commented:
Only a few small things, otherwise looks good! 

Tab hover - Set both the font and border to #404040.
Next PAGE arrow - Icon should be 12x24px. The total height of the component should be 30px.
Add a Task prompt - Font size should be 18pt, icon 12px. The total height of the component should be 30px.

We made a note in our discussion to revisit the drop target in the future.

shaunanoordin commented:
Dangit, I forgot to add frontenddevs as reviewers last week. I've just fixed that, and I'm also making some small CSS updates based on [Sean's recent comments.](https://github.com/zooniverse/Panoptes-Front-End/pull/7245#issuecomment-2603287583)

shaunanoordin commented:
## PR Update

I've addressed [Sean's feedback.](https://github.com/zooniverse/Panoptes-Front-End/pull/7245#issuecomment-2603287583) 👌 

- Workflow Header Tabs, on _hover state,_ now have a consistent #404040 colour.
- "Next Task Arrow" now uses SVG as specified in Figma, instead of Font Awesome placeholders, so more accurately follows the dimensions stated.
- "+" decoration in "Add A New Task +" button is now an actual icon instead of a CSS decoration, and also used the SVG specified in Figma.

ALSO:
- I've updated the README with some notes about the naming convention of the Pages Editor project. 🐱 Read it meow!

kieftrav reviewed (approved):
Huge fan of the refactor to component-based classes. Makes inspecting elements much easier and following CSS easier as well! Didn't focus on anything design-wise as I saw the copious feedback from Sean. I focused instead on the called-out developer review actions (creating/editing/deleting tasks + pages), of which everything worked and updated as expected. All code comments are more style/pattern based to match the standards I have been receiving in my code reviews. I'm approving but the only thing that needs to be resolved before merging is the 1 failing test.

shaunanoordin commented on app/pages/lab-pages-editor/README.md:
```diff
@@ -1,9 +1,20 @@
+A note about the naming:
+- When this project started in late 2023, the new UI for editing workflows was called "Pages Editor", because it was a workflow editor that organised workflows into discrete pages (steps).
+- In late 2024, the project name was changed to "Workflow Editor". This is the equivalent of renaming your third cat, who was originally called "Kitty" to differentiate her from older siblings "Feline" and "Feline 2", to simply "Cat".
+- To (hopefully?) minimuse confusion, the code itself will continue to use "Pages Editor" naming scheme, e.g. `lab-pages-editor.styl`
```
Me fail English? That's unpossible!

shaunanoordin commented on app/pages/lab-pages-editor/components/TasksPage/components/StepItem/NextStepControls.jsx:
```diff
@@ -0,0 +1,79 @@
+  
+  function onChange(e) {
+    const next = e.target?.value;
+    updateNextStepForStep(stepKey, next);
```
- It's entirely possible for `step.next` to be `undefined`, so this is OK.
  - IIRC, if `step.next = ''` or `step.next = undefined` that means "the next step is to Submit".
  - Unfortunately, I'm getting weird issues testing the workflow on staging at the moment, so I've added "confirm what I said isn't bullcrap" on my next week's TODO list. 🤔 
- I've updated TaskPage's updateNextStepForStep() with slightly better documentation for its expected inputs. i.e. "stepKey" has to be valid, but "next" is optional.
  - see https://github.com/zooniverse/Panoptes-Front-End/pull/7245/commits/ac0d0ce80ca14bae57e03a4563716bc664b162e7
  - I've also added additional TODOs here, since updateNextStepForStep() should check for valid input. That's a lotta TODOs! 😬 
- Fortunately, the "manually change Next Step" is a feature hidden behind a hidden "advanced settings" option (add `?advanced=1` to the URL) so this isn't critical to fix for most users. 👌

shaunanoordin replied:
Update update: confirmed, on the FEM Classifier, if `step.next = undefined` or `step.next = ''`, then the "next step" is to submit the Classification.

shaunanoordin commented on app/pages/lab-pages-editor/components/TasksPage/components/StepItem/NextStepControls.jsx:
```diff
@@ -0,0 +1,79 @@
+              Submit
+            </option>
+            {selectableSteps.map(([otherStepKey, otherStepBody]) => {
+              const taskKeys = otherStepBody?.taskKeys?.toString() || '(none)';
```
Good point here.

- I've actually renamed it to `taskKeysAsLabel` - still plural, but better explains what this is.
- If a Step S0 has tasks [ T0, T1, T2 ], then the display label/display name for the step isn't "S0", but actually "T0, T1, T2"

Example: The label for the second page (internally identified as step.id = P2) is displayed as "T2, T3" in the yellow/orange selector.
<img width="400" alt="image" src="https://github.com/user-attachments/assets/6c985b63-eaf9-42ed-a04a-c4eb47e18a71" />

Updated in https://github.com/zooniverse/Panoptes-Front-End/pull/7245/commits/04e73b4084b83c4fb28200ef7b069f6e39fd97c7

shaunanoordin commented on app/pages/lab-pages-editor/components/WorkflowHeaderTabs.jsx:
```diff
@@ -47,46 +38,29 @@ export default function WorkflowHeader({
+        <TabButton
+          id={tab.id}
+          index={index}
+          key={`${tab.id}`}
```
Also, experimental trivia: it appears that just using a number for a key (i.e. `key={tab.id}`) seems to work as well. I'm surprised React didn't throw more of a fuss!

shaunanoordin commented on app/pages/lab-pages-editor/components/WorkflowHeaderTitle.jsx:
```diff
@@ -0,0 +1,28 @@
+
+  if (!project || !workflow) return null;
+
+  const workflowTitle = workflow?.display_name ?? '(unnamed workflow)';
```
The `?.` conditionals are unnecessary here actually, since I start this function with `if (!project || !workflow) return null`, but I tend to code very defensively by default. One might say to the point of mild paranoia!

shaunanoordin commented on app/pages/lab-pages-editor/components/WorkflowTopLinks.jsx:
```diff
@@ -0,0 +1,25 @@
+
+export default function WorkflowTopLinks() {
+  const { project, workflow } = useWorkflowContext();
+  const returnUrl = `/lab/${project?.id}/workflows`;
```
Ditto! ☝️

shaunanoordin commented on app/pages/lab-pages-editor/components/WorkflowHeaderTitle.jsx:
```diff
@@ -0,0 +1,28 @@
+
+  const workflowTitle = workflow?.display_name ?? '(unnamed workflow)';
+  const previewEnv = getPreviewEnv();
+  const previewUrl = `https://frontend.preview.zooniverse.org/projects/${project?.slug}/classify/workflow/${workflow?.id}${previewEnv}`;
```
Ditto! ☝️

shaunanoordin commented on app/pages/lab-pages-editor/icons/NextStepIcon.jsx:
```diff
@@ -0,0 +1,20 @@
+export default function NextStepIcon({ alt, className = '', width = 30, height = 30 }) {
```
Honestly, I'm still figuring this one out - I actually had default alts at one point! But then I learnt that _purely decorative images_ probably shouldn't have alt text (or rather, should have `alt=""`) to prevent noise from the voice overs.

https://www.boia.org/blog/when-should-i-use-empty-or-null-alt-text-attributes

For now, I try to leave the icons with blank alts by default, with the responsibility of specifying alt to whatever component's including those icons.

(The rules are slightly different for SVGs though, but your comment here did remind me that I forgot to add `<svg role="img">` to the new icons. Fixed!)

shaunanoordin commented:
Thanks for the thorough review, Travis! 👍 I've addressed all the feedback you've provided, except for the "de-structure PropTypes" one - I'm going to need to better understand what Delilah did first, and then apply it to _every_ component in `/lab-pages-editor` for consistency. So that's probably going to be its own refactor PR! (Also that PR should clean up a bunch es-lint exceptions I baked into the code.)

For now, let's get this merged! Let's gooo

goplayoutside3 commented:
@kieftrav destructuring applies to [React default values](https://react.dev/learn/passing-props-to-a-component#specifying-a-default-value-for-a-prop) for props in functional components. It does not apply to generalized PropTypes. The PropTypes library has a `.defaultProps` feature, but that feature is not recognized by React v19, so code in FEM is moving toward destructured _default values_ instead.

goplayoutside3 commented on app/pages/lab-pages-editor/components/TasksPage/components/StepItem/NextStepControls.jsx:
```diff
@@ -0,0 +1,79 @@
+}
+
+NextStepControls.propTypes = {
+  allSteps: PropTypes.arrayOf(PropTypes.array),
```
@shaunanoordin I was using 'destructuring in the import statement' over in FEM and realized why there's confusion in this PR review. Travis is referring to this syntax (please correct me if I'm wrong):
`import { bool, func } from 'prop-types'` instead of `import Proptypes from 'prop-types'` for the purpose of tree-shaking. I'm not sure how well tree-shaking works in PFE's app, but thought I'd clear up any confusion based on your comment post-review:
> I'm going to need to better understand what Delilah did first, and then apply it to every component
